### PR TITLE
Update things such that they can be built and deployed with Bazel.

### PR DESCRIPTION
**Pull request by mattmoor**
To configure whether to publish images and which cluster to target:
```shell
# Put this in .bashrc (or w/e)
export DOCKER_REPO_OVERRIDE=us.gcr.io/project
export K8S_CLUSTER_OVERRIDE=cluster-name
```

Register the CRD with:
```shell
bazel run artifacts/examples:crd.create
```

Then build/push/deploy the controller to your cluster:
```shell
bazel run :controller.create
```

Test the controller with:
```shell
bazel run artifacts/examples:example.create
```

Iterate on changes to the controller (on-cluster) with:
```shell
bazel run :controller.replace
```

### CODE REVIEW

File: Gopkg.lock [hunk 1]
```diff
@@ -432,28 +433,9 @@
   ]
   revision = "39a7bf85c140f972372c2a0d1ee40adbf0c8bfe1"
 
-[[projects]]
```
Comment (mattmoor) [inline Gopkg.lock]: I think this difference stems from you doing things on a copy created from your repo, whereas I did this on `k8s.io/sample-controller` and then added my fork as a new remote (so this isn't needed because it's us!).  If we want to call this something other than `k8s.io/sample-controller`, then I think we need to rewrite references so that we don't have a dependency on it.

**Comment by mattmoor** [comment]
So the rough process I used to update deps was:
1. `dep ensure` (with a fake dep in `controller.go` to get the `k8s.io/apimachinery/pkg/util/sets/types` directory vendored)
1. `bazel run //:gazelle -- -proto=disable` to generate the `BUILD` files
1. The following command to rewrite a Bazel dependency that doesn't vendor well:
```shell
# Turn references to k8s.io/code-generator under vendor/ into Bazel repository references.
sed -i 's|//vendor/k8s.io/code-generator/|@io_k8s_code_generator//|g' \
   $(find -name '*' | xargs grep k8s.io/code-generator |& grep code-gen | cut -d':' -f 1 | uniq)
```

**Comment by mattmoor** [comment]
@ixdy @ImJasonH FYI

File: Gopkg.lock (hunk 1 above)
Comment (grantr) [reply Gopkg.lock]: Just to confirm I understand: my process was

```sh
(fork)
dep init
```

and your process was
```sh
dep init
(fork)
```
> If we want to call this something other than k8s.io/sample-controller, then I think we need to rewrite references so that we don't have a dependency on it.

:+1: I expect `dep` will gracefully detect future package renames.

**Comment by grantr** [comment]
@mattmoor did you run `dep prune` after `dep ensure`? I normally prefer that because it reduces repo and diff size.

File: Gopkg.lock (hunk 1 above)
Comment (mattmoor) [reply Gopkg.lock]: I never called `dep init`, since I think that's one-time setup.

What I did was:
```shell
go get k8s.io/sample-controller
cd $GOPATH/src/k8s.io/sample-controller
git remote add mattmoor <my fork>
./hack/update-deps.sh
```

So my point was that I'm working within a directory named `k8s.io/sample-controller` on my `GOPATH`.

**Comment by mattmoor** [comment]
@grantr I ran what's in `./hack/update-deps.sh`, so no.  I'll add that there and update the PR.

**Comment by mattmoor** [comment]
@grantr should be RFAL.

File: Gopkg.lock (hunk 1 above)
Comment (grantr) [reply Gopkg.lock]: > So my point was that I'm working within a directory named `k8s.io/sample-controller` on my `GOPATH`.

Do we expect this fork to ultimately be merged back into the parent? If not I think it makes sense to choose a new prefix.

File: controller.go [hunk 2]
```diff
@@ -33,6 +33,9 @@ import (
 	"k8s.io/client-go/tools/record"
 	"k8s.io/client-go/util/workqueue"
 
+	// Trick dep (DO THIS BETTER)
+	_ "k8s.io/apimachinery/pkg/util/sets/types"
```
Comment (grantr) [inline controller.go]: For my edification, why is this needed?

File: Gopkg.lock (hunk 1 above)
Comment (mattmoor) [reply Gopkg.lock]: I doubt it.  I think this PR makes it *slightly* more correct (since it isn't depending on vendored code, but it's own copy), but I think a subsequent commit that `s@k8s.io/sample-controller@foo.bar/baz@g` would be fine.

File: controller.go (hunk 2 above)
Comment (mattmoor) [reply controller.go]: It isn't strictly needed **here** (and should probably move elsewhere so it isn't being linked), but we need it *somewhere* to trick `dep` into venodring this particular sub-portion of that repo (it's needed by Bazel).

Comment (mattmoor) [reply controller.go]: I backed this out, just to grab the location here it breaks, see [here](https://github.com/mattmoor/sample-controller/blob/bbae7b8ebb613ede9bd444e2700ee12dcedf9cf6/vendor/k8s.io/apimachinery/pkg/util/sets/BUILD#L23-L49)

Comment (grantr) [reply controller.go]: I think we can do that in dep by adding a `required` directive to `Gopkg.toml`: https://github.com/golang/dep/blob/master/docs/Gopkg.toml.md

```toml
# needed by Bazel
required = ["k8s.io/apimachinery/pkg/util/sets/types"]
```

File: Gopkg.lock (hunk 1 above)
Comment (grantr) [reply Gopkg.lock]: Just tested it; changing `go_prefix` and rerunning `hack/update-deps.sh` was enough to update everything.

File: vendor/k8s.io/client-go/util/cert/BUILD [hunk 3]
```diff
@@ -25,9 +25,10 @@ go_library(
         "io.go",
         "pem.go",
     ],
-    data = [
-        "testdata/dontUseThisKey.pem",
-    ],
+    # Hand editted out, as this doesn't seem to be required.
```
Comment (grantr) [inline vendor/k8s.io/client-go/util/cert/BUILD]: This change was overwritten by `hack/update-deps.sh`, so if it needs to stick around it should be automated.

File: controller.go (hunk 2 above)
Comment (grantr) [reply controller.go]: Submitted a PR to your fork at https://github.com/mattmoor/sample-controller/pull/1

File: vendor/k8s.io/client-go/util/cert/BUILD (hunk 3 above)
Comment (mattmoor) [reply vendor/k8s.io/client-go/util/cert/BUILD]: Done, this is changed in `./hack/update-deps.sh` now.
